tests/test_pandas_utils/test_pandas_utils.py

```diff
@@ -73,8 +72,260 @@ def do_nothing(df, *args, **kwargs):
+
+    assert ("dtypes=" in captured.out) == dtypes
+
+    if dtypes:
+        assert str(test_df.dtypes.to_dict()) in captured.out
+
+
+def test_log_not_names_and_dtypes(capsys, test_df):
+
+    @log_step(names=True, dtypes=True)
+    def do_nothing(df, *args, **kwargs):
+        return df
+
+    test_df.pipe(do_nothing)
+
+    captured = capsys.readouterr()
+
+    assert "names=" not in captured.out
+
+
+def test_log_custom_logger(caplog, test_df):
+    caplog.clear()
+
+    logger_name = "my_custom_logger"
+
+    my_logger = logging.getLogger(logger_name)
+
+    @log_step(print_fn=my_logger.info)
+    def do_nothing(df, *args, **kwargs):
+        return df
+
+    test_df.pipe(do_nothing)
+
+    assert logger_name in caplog.text
+
+
+def test_log_extra(capsys):
+
+    n_cats = 3
+    n_dogs = 2
+
+    test_df = pd.DataFrame({
+        "id": range(n_cats + n_dogs),
+        "animals": ["dog"] * n_dogs + ["cat"] * n_cats,
+    })
+
+    def cat_counter(df):
+        return f"cats={(df['animals']=='cat').sum()}"
+
+    @log_step_extra(cat_counter)
+    def do_nothing(df, *args, **kwargs):
+        return df
+
+    @log_step_extra(cat_counter)
+    def double_df(df, *args, **kwargs):
+        return pd.concat([df, df], axis=0)
+
+    test_df.pipe(do_nothing).pipe(double_df)
+
+    captured = capsys.readouterr()
+    print_statements = captured.out.split("\n")
+
+    assert f"cats={n_cats}" in print_statements[0]
+    assert f"cats={2*n_cats}" in print_statements[1]
+
+
+def test_log_extra_kwargs(capsys):
+
+    n_cats = 3
+    n_dogs = 2
+
+    test_df = pd.DataFrame({
+        "id": range(n_cats + n_dogs),
+        "animals": ["dog"] * n_dogs + ["cat"] * n_cats,
+    })
+
+    def animal_counter(df, animal="cat"):
+        return f"{animal}s={(df['animals']==animal).sum()}"
+
+    @log_step_extra(animal_counter, animal="dog")
+    def do_nothing(df, *args, **kwargs):
+        return df
+
+    @log_step_extra(animal_counter, animal="dog")
+    def double_df(df, *args, **kwargs):
+        return pd.concat([df, df], axis=0)
+
+    test_df.pipe(do_nothing).pipe(double_df)
+
+    captured = capsys.readouterr()
+    print_statements = captured.out.split("\n")
+
+    assert f"dogs={n_dogs}" in print_statements[0]
+    assert f"dogs={2*n_dogs}" in print_statements[1]
+
+
+def test_log_extra_multiple(capsys, test_df):
+
+    @log_step_extra(len, type)
+    def do_nothing(df, *args, **kwargs):
+        return df
+
+    test_df.pipe(do_nothing)
+
+    captured = capsys.readouterr()
+
+    assert str(len(test_df)) in captured.out
+    assert str(type(test_df)) in captured.out
+
+
+def test_log_extra_no_func(test_df):
+    with pytest.raises(ValueError) as e:
+        @log_step_extra()
+        def do_nothing(df, *args, **kwargs):
+            return df
+
+        test_df.pipe(do_nothing)
+
+        assert "log_function" in str(e)
+
+
+def test_log_extra_not_callable_func(test_df):
+    with pytest.raises(ValueError) as e:
+        @log_step_extra(1)
```

Comment:
It might be nice to either add a docstring or a comment explaining what we're testing with these unit tests. Our memory is still fresh now but a year from now somebody might like to read "you must pass a callable to `log_step_extra`" in this test.

Reply:
Done!